### ISSUE TITLE
Creator and type information for Deferreds

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5722]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#5722|
|Type|enhancement|
|Created|2012-06-26 20:44:04Z|

When you get a Deferred, you will typically see something like `<Deferred at 0x10a541f38>`.

But when you read the documentation for the thing that gave you that `Deferred`, it will tell you something about what the `Deferred` fires with.

I'd much rather see something like `<Deferred originated by 'twisted.web.client.getPage', which will fire with 'int', at 0x10a541f38>`.

In order to provide this information, callbacks could be annotated with a return type, and `Deferred.__init__` would be modified to allow providing the name of the originator.  `inlineCallbacks` would need special support.

This state could also be provided for introspection and debugging aids, similar to how Zope Interface provides us with tools to interrogate the type and claimed interface compatibility of an object.

<details><summary>Searchable metadata</summary>

```
trac-id__5722 5722
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__new new
resolution__None None
component__core core
keywords__ 
time__1340743444000000 1340743444000000
changetime__1560836528674961 1560836528674961
version__None None
owner__None None

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to @glyph|
|-|-|

An "and" in a ticket summary is a common sign that a ticket is too broadly scoped.

  * As far as the originator goes, how useful is it to see `twisted.web.client.getPage` if you're using `twisted.web.xmlrpc.Proxy`?  I called `callRemote`.  Seeing `getPage` doesn't reveal that to me, and is potentially actively misleading.
  * As far as the return type information goes, every callback is a new chance for the types related to the Deferred to change.  You can't annotate the Deferred with a return type by passing an `__init__` argument.  You would have to annotate it at `addCallback` time.

I think this needs some more thought and specification.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

Also, don't forget about [#4303](https://github.com/twisted/twisted/issues/4303), which has had some development effort directed at it already.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_1):
> An "and" in a ticket summary is a common sign that a ticket is too broadly scoped.

I fully expect that this ticket will spawn a few others.  I intentionally called it "RTTI" because I would like it to have some umbrella scope, possibly not just return type, but other things about the callback chain.

>   * As far as the originator goes, how useful is it to see `twisted.web.client.getPage` if you're using `twisted.web.xmlrpc.Proxy`?  I called `callRemote`.  Seeing `getPage` doesn't reveal that to me, and is potentially actively misleading.

So perhaps the "originator" should also be the most recently added callback as well?  So each level informs the next of what the Deferred's current purpose is?

>   * As far as the return type information goes, every callback is a new chance for the types related to the Deferred to change.  You can't annotate the Deferred with a return type by passing an `__init__` argument.  You would have to annotate it at `addCallback` time.

Oh, um. Yes, absolutely. I was thinking that the `__init__` argument could say something about the creator, since that is where e.g. the canceler is also set up.  And perhaps the ultimate causality is interesting sometimes.  This is less about the "originator" though and more about who you expect to _fire_ the Deferred.  This would change as you had callbacks that returned other `Deferred`s as well.  Would 'waiting on' be better language here?

> I think this needs some more thought and specification.

Absolutely. That was the intent. Is there a keyword or something I should use to indicate tickets in this state?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

> Absolutely. That was the intent. Is there a keyword or something I should use to indicate tickets in this state?

Maybe this would be better as a wiki page ([eg](https://twistedmatrix.com/trac/wiki/Specification/AsynchronousFileInputOutput)).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph removed owner|
|-|-|
